File: Editorial/115_Distinct_Subsequences/TypeScript/1_Recursion_Memoization.ts
function numDistinct(s: string, t: string): number {
    let memo = new Map<string, number>();
    function helper(i: number, j: number): number {
        if (i === s.length || j === t.length || s.length - i < t.length - j)
            return j === t.length ? 1 : 0;
        let key = i + "," + j;
        if (memo.has(key)) return memo.get(key) as number;
        let ans = helper(i + 1, j);
        if (s[i] === t[j]) ans += helper(i + 1, j + 1);
        memo.set(key, ans);
        return ans;
    }
    return helper(0, 0);
}

File: Editorial/115_Distinct_Subsequences/Go/1_Recursion_Memoization.go
func numDistinct(s string, t string) int {
    memo := make(map[string]int)
    var helper func(i int, j int) int
    helper = func(i int, j int) int {
        if i == len(s) || j == len(t) || len(s)-i < len(t)-j {
            if j == len(t) {
                return 1
            }
            return 0
        }
        key := strconv.Itoa(i) + "," + strconv.Itoa(j)
        if val, ok := memo[key]; ok {
            return val
        }
        ans := helper(i+1, j)
        if s[i] == t[j] {
            ans += helper(i+1, j+1)
        }
        memo[key] = ans
        return ans
    }
    return helper(0, 0)
}
